File: src/main/java/slavsquatsuperstar/mayonez/physics2d/colliders/Edge2D.java
package slavsquatsuperstar.mayonez.physics2d.colliders;

import slavsquatsuperstar.math.MathUtils;
import slavsquatsuperstar.math.Vec2;
import slavsquatsuperstar.mayonez.physics2d.RaycastResult;

/**
 * A 2D line segment with start and end points.
 *
 * @author SlavSquatSuperstar
 */
public class Edge2D {

    public final Vec2 start, end;

    public Edge2D(Vec2 start, Vec2 end) {
        this.start = start;
        this.end = end;
    }

    // Properties

    float getSlope() {
        return (end.y - start.y) / (end.x - start.x);
    }

    public float length() {
        return start.distance(end);
    }

    public Vec2 toVector() {
        return end.sub(start);
    }

    // Line vs Point

    /**
     * Calculates whether the given point is on this line segment.
     *
     * @param point a point in 2D space
     * @return if the point is on the line
     */
    public boolean contains(Vec2 point) {
        if (point.equals(start) || point.equals(end))
            return true;
        return nearestPoint(point).equals(point);
    }

    public Vec2 nearestPoint(Vec2 position) {
        float length = length();
        float projLength = position.sub(start).dot(toVector()) / length; // find point shadow on line
        if (projLength > length) // pat line end
            return end;
        else if (projLength < 0) // behind line start
            return start;
        return start.add(toVector().mul(projLength / length)); // inside line
    }

    // right (clockwise) is positive
    public float distance(Vec2 point) {
        return point.sub(start).projectedLength(toVector().rotate(90));
    }

    /**
     * Returns the distance from a point to this line relative to that of other points, and which side the point is on.
     *
     * @param point a point
     * @return A negative value if the point is to the right of this line, a positive value if the point is to the left
     * of this line, and zero if the point is on the line (if extended infinitely).
     */
    public float pseudoDistance(Vec2 point) {
        return point.sub(start).dot(toVector().rotate(90));
    }

    // Line vs Line

    public boolean intersects(Edge2D edge) {
        // If parallel, must be overlapping (co-linear)
        if (MathUtils.equals(this.getSlope(), edge.getSlope())) {
            return this.contains(edge.start) || this.contains(edge.end) ||
                    edge.contains(this.start) || edge.contains(this.end);
        }
        return raycast(new Ray2D(edge), edge.length()) != null;
    }

    public RaycastResult raycast(Ray2D ray, float limit) {
        float length = this.length();
        Vec2 start1 = this.start;
        Vec2 start2 = ray.origin;

        // Calculate point of intersection using cross product
        Vec2 line1 = this.toVector().div(length);
        Vec2 line2 = ray.direction;
        float cross = line1.cross(line2);

        // Parametric lengths along rays
        float dist1 = start2.sub(start1).cross(line2) / cross;
        float dist2 = start1.sub(start2).cross(line1) / -cross;

        if (!MathUtils.inRange(dist1, 0, length)) // Contact point outside line
            return null;
        if (limit > 0 && !MathUtils.inRange(dist2, 0, limit)) // Contact point exceeds ray limit
            return null;
        else if (dist2 < 0) // Contact point behind ray
            return null;

        Vec2 contact = start1.add(line1.mul(dist1));
        // rotate left or right depending on which side ray started form
        Vec2 normal = line1.rotate(Math.signum(pseudoDistance(start2)) * 90);
        return new RaycastResult(contact, normal, dist1);
    }

    @Override
    public String toString() {
        return String.format("Start: %s, End: %s", start, end);
    }
}

File: src/main/kotlin/slavsquatsuperstar/mayonez/physics2d/colliders/Edge2D.kt
package slavsquatsuperstar.mayonez.physics2d.colliders

import slavsquatsuperstar.math.MathUtils
import slavsquatsuperstar.math.Vec2
import slavsquatsuperstar.mayonez.physics2d.RaycastResult
import kotlin.math.sign

/**
 * A 2D line segment with start and end points.
 *
 * @author SlavSquatSuperstar
 */
class Edge2D(@JvmField val start: Vec2, @JvmField val end: Vec2) {

    // Properties
    fun slope(): Float = (end.y - start.y) / (end.x - start.x)

    fun length(): Float = start.distance(end)

    fun toVector(): Vec2 = end.sub(start)

    // Line vs Point
    /**
     * Calculates whether the given point is on this line segment.
     *
     * @param point a point in 2D space
     * @return if the point is on the line
     */
    operator fun contains(point: Vec2): Boolean {
        return if (point == start || point == end) true
        else nearestPoint(point) == point
    }

    fun nearestPoint(position: Vec2): Vec2 {
        val length = length()
        val projLength = position.sub(start).dot(toVector()) / length // find point shadow on line
        if (projLength > length) // pat line end
            return end else if (projLength < 0) // behind line start
            return start
        return start.add(toVector().mul(projLength / length)) // inside line
    }

    // right (clockwise) is positive
    fun distance(point: Vec2): Float {
        return point.sub(start).projectedLength(toVector().rotate(90f))
    }

    /**
     * Returns the distance from a point to this line relative to that of other points, and which side the point is on.
     *
     * @param point a point
     * @return A negative value if the point is to the right of this line, a positive value if the point is to the left
     * of this line, and zero if the point is on the line (if extended infinitely).
     */
    fun pseudoDistance(point: Vec2): Float {
        return point.sub(start).dot(toVector().rotate(90f))
    }

    // Line vs Line
    fun intersects(edge: Edge2D): Boolean {
        // If parallel, must be overlapping (co-linear)
        return if (MathUtils.equals(slope(), edge.slope())) {
            this.contains(edge.start) || this.contains(edge.end) ||
                    edge.contains(start) || edge.contains(end)
        } else raycast(Ray2D(edge), edge.length()) != null
    }

    fun raycast(ray: Ray2D, limit: Float): RaycastResult? {
        val length = length()
        val start1 = start
        val start2 = ray.origin

        // Calculate point of intersection using cross product
        val line1 = toVector().div(length)
        val line2 = ray.direction
        val cross = line1.cross(line2)

        // Parametric lengths along rays
        val dist1 = start2.sub(start1).cross(line2) / cross
        val dist2 = start1.sub(start2).cross(line1) / -cross
        if (!MathUtils.inRange(dist1, 0f, length)) // Contact point outside line
            return null
        if (dist2 < 0 || limit > 0 && dist2 > limit) // Contact point outside allowed ray
            return null
        val contact = start1.add(line1.mul(dist1))
        // rotate left or right depending on which side ray started form
        val normal = line1.rotate(sign(pseudoDistance(start2)) * 90)
        return RaycastResult(contact, normal, dist1)
    }

    override fun toString(): String = String.format("Start: %s, End: %s", start, end)
}